File: knowledge_graph/test/test_edge.cpp
// Copyright 2025 Miguel Ángel González Santamarta
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <gtest/gtest.h>

#include "knowledge_graph/graph/edge.hpp"
#include "knowledge_graph_msgs/msg/content.hpp"
#include "knowledge_graph_msgs/msg/edge.hpp"
#include "knowledge_graph_msgs/msg/property.hpp"

using namespace knowledge_graph::graph;
using knowledge_graph_msgs::msg::Content;
using knowledge_graph_msgs::msg::Property;

class EdgeTest : public ::testing::Test {};

// Constructor Tests
TEST_F(EdgeTest, ConstructorWithParameters) {
  Edge edge("connects", "node_a", "node_b");
  EXPECT_EQ(edge.get_type(), "connects");
  EXPECT_EQ(edge.get_source_node(), "node_a");
  EXPECT_EQ(edge.get_target_node(), "node_b");
}

TEST_F(EdgeTest, ConstructorFromMessage) {
  knowledge_graph_msgs::msg::Edge msg;
  msg.type = "relates";
  msg.source_node = "source";
  msg.target_node = "target";

  Edge edge(msg);
  EXPECT_EQ(edge.get_type(), "relates");
  EXPECT_EQ(edge.get_source_node(), "source");
  EXPECT_EQ(edge.get_target_node(), "target");
}

TEST_F(EdgeTest, ConstructorFromMessageWithProperties) {
  Property prop;
  prop.key = "weight";
  prop.value.type = Content::DOUBLE;
  prop.value.double_value = 0.75;

  knowledge_graph_msgs::msg::Edge msg;
  msg.type = "connects";
  msg.source_node = "a";
  msg.target_node = "b";
  msg.properties.push_back(prop);

  Edge edge(msg);
  EXPECT_EQ(edge.get_type(), "connects");
  EXPECT_TRUE(edge.has_property("weight"));
  EXPECT_NEAR(edge.get_property<double>("weight"), 0.75, 0.001);
}

// Message Conversion Tests
TEST_F(EdgeTest, ToMsg) {
  Edge edge("connects", "node_a", "node_b");
  edge.set_property<double>("weight", 1.5);

  auto msg = edge.to_msg();
  EXPECT_EQ(msg.type, "connects");
  EXPECT_EQ(msg.source_node, "node_a");
  EXPECT_EQ(msg.target_node, "node_b");
  EXPECT_EQ(msg.properties.size(), 1u);
  EXPECT_EQ(msg.properties[0].key, "weight");
}

// String Representation Tests
TEST_F(EdgeTest, ToString) {
  Edge edge("connects", "node_a", "node_b");
  auto str = edge.to_string();
  EXPECT_NE(str.find("connects"), std::string::npos);
  EXPECT_NE(str.find("node_a"), std::string::npos);
  EXPECT_NE(str.find("node_b"), std::string::npos);
}

// Property Tests
TEST_F(EdgeTest, SetAndGetProperty) {
  Edge edge("connects", "node_a", "node_b");
  edge.set_property<double>("weight", 0.9);
  edge.set_property<std::string>("label", "connection");
  edge.set_property<bool>("bidirectional", true);

  EXPECT_NEAR(edge.get_property<double>("weight"), 0.9, 0.001);
  EXPECT_EQ(edge.get_property<std::string>("label"), "connection");
  EXPECT_TRUE(edge.get_property<bool>("bidirectional"));
}

TEST_F(EdgeTest, HasProperty) {
  Edge edge("connects", "node_a", "node_b");
  EXPECT_FALSE(edge.has_property("weight"));
  edge.set_property<double>("weight", 0.5);
  EXPECT_TRUE(edge.has_property("weight"));
}

TEST_F(EdgeTest, PropertyRoundtrip) {
  Edge edge("connects", "node_a", "node_b");
  edge.set_property<double>("weight", 0.75);
  edge.set_property<std::string>("label", "test_edge");
  edge.set_property<bool>("enabled", false);

  auto msg = edge.to_msg();
  Edge restored_edge(msg);

  EXPECT_EQ(restored_edge.get_type(), "connects");
  EXPECT_EQ(restored_edge.get_source_node(), "node_a");
  EXPECT_EQ(restored_edge.get_target_node(), "node_b");
  EXPECT_NEAR(restored_edge.get_property<double>("weight"), 0.75, 0.001);
  EXPECT_EQ(restored_edge.get_property<std::string>("label"), "test_edge");
  EXPECT_FALSE(restored_edge.get_property<bool>("enabled"));
}

int main(int argc, char **argv) {
  testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}

File: knowledge_graph/test/test_edge.py
# Copyright 2025 Miguel Ángel González Santamarta

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at

#     http://www.apache.org/licenses/LICENSE-2.0

# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from knowledge_graph.graph.edge import Edge
from knowledge_graph_msgs.msg import Edge as EdgeMsg, Property, Content


class TestEdge:
    """Test suite for Edge class."""

    def test_constructor_with_parameters(self):
        """Test constructor with type, source, and target."""
        edge = Edge("connects", "node_a", "node_b")
        assert edge.get_type() == "connects"
        assert edge.get_source_node() == "node_a"
        assert edge.get_target_node() == "node_b"

    def test_constructor_from_message(self):
        """Test constructor from ROS message."""
        msg = EdgeMsg()
        msg.type = "relates"
        msg.source_node = "source"
        msg.target_node = "target"

        edge = Edge(msg=msg)
        assert edge.get_type() == "relates"
        assert edge.get_source_node() == "source"
        assert edge.get_target_node() == "target"

    def test_constructor_from_message_with_properties(self):
        """Test constructor from message with properties."""
        prop = Property()
        prop.key = "weight"
        prop.value.type = Content.DOUBLE
        prop.value.double_value = 0.75

        msg = EdgeMsg()
        msg.type = "connects"
        msg.source_node = "a"
        msg.target_node = "b"
        msg.properties = [prop]

        edge = Edge(msg=msg)
        assert edge.get_type() == "connects"
        assert edge.has_property("weight")
        assert abs(edge.get_property("weight") - 0.75) < 0.001

    def test_to_msg(self):
        """Test converting Edge to ROS message."""
        edge = Edge("connects", "node_a", "node_b")
        edge.set_property("weight", 1.5)

        msg = edge.to_msg()
        assert msg.type == "connects"
        assert msg.source_node == "node_a"
        assert msg.target_node == "node_b"
        assert len(msg.properties) == 1
        assert msg.properties[0].key == "weight"

    def test_to_string(self):
        """Test string representation of Edge."""
        edge = Edge("connects", "node_a", "node_b")
        string_repr = edge.to_string()
        assert "connects" in string_repr
        assert "node_a" in string_repr
        assert "node_b" in string_repr

    def test_str_method(self):
        """Test __str__ method."""
        edge = Edge("connects", "node_a", "node_b")
        assert str(edge) == edge.to_string()

    def test_set_and_get_property(self):
        """Test setting and getting properties on an edge."""
        edge = Edge("connects", "node_a", "node_b")
        edge.set_property("weight", 0.9)
        edge.set_property("label", "connection")
        edge.set_property("bidirectional", True)

        assert abs(edge.get_property("weight") - 0.9) < 0.001
        assert edge.get_property("label") == "connection"
        assert edge.get_property("bidirectional") is True

    def test_has_property(self):
        """Test checking if a property exists."""
        edge = Edge("connects", "node_a", "node_b")
        assert not edge.has_property("weight")
        edge.set_property("weight", 0.5)
        assert edge.has_property("weight")

    def test_property_roundtrip(self):
        """Test that properties survive a message roundtrip."""
        edge = Edge("connects", "node_a", "node_b")
        edge.set_property("weight", 0.75)
        edge.set_property("label", "test_edge")
        edge.set_property("enabled", False)

        msg = edge.to_msg()
        restored_edge = Edge(msg=msg)

        assert restored_edge.get_type() == "connects"
        assert restored_edge.get_source_node() == "node_a"
        assert restored_edge.get_target_node() == "node_b"
        assert abs(restored_edge.get_property("weight") - 0.75) < 0.001
        assert restored_edge.get_property("label") == "test_edge"
        assert restored_edge.get_property("enabled") is False
